File: tests/testDataFieldInherit.java
/*Resolve Data Fields:
-After Structure Parser Completes
1. Pass assignment node to eval call to see what the dynamic type will be
2. Need to find the correct scope in the symbol table and add it

Then, inherit datafields
1.
*/

public class testDataFieldInherit{

    String name = "original name";

    public static void main(String[] args)
    {
        testI1 t1 = new testI1();
        t1.test();
        t1.test2();

        testI2 t2 = new testI2();
        t2.test();
        t2.test2();
		testI3 t3 = new testI3();
		t3.test();

		testI4 t4 = new testI4();
		t4.test();
    }

}

class testI1 extends testDataFieldInherit{
	//vtable:
	//string name;
    public void test(){
        System.out.println("test1.test(): Should print original name:");
        System.out.println("\t" + this.name);
    }

    public void test2(){
        System.out.println("test1.test2(): Should print original name:");
        System.out.println("\t" + super.name);
    }
}

class testI2 extends testDataFieldInherit{
	//vtable:
	//String $name;
	//int name
    int name = 69;

    public void test(){
        System.out.println("test2.test(): Should print 69:");
        System.out.println("\t" + name);
    }

    public void test2(){
        System.out.println("test2.test2(): Should print original name:");
        System.out.println("\t" + super.name);
    }


}

class testI3 extends testI2{
	//vtable
	//String $$name
	//int $name
	//char name
	char name = 'c';

	public void test(){
		System.out.println("Should print 69: " + super.name);
		System.out.println("Shoudld print c: " + name);
	}

}

class testI4 extends testI3{

	public void test(){
		System.out.println("Should print c: " + super.name);
		System.out.println("Shoudld print c: " + name);

	}
}

File: tests/translated/testDataFieldInherit.cc
#include "testDataFieldInherit.h"
#include <iostream>

void _testDataFieldInherit::__delete(_testDataFieldInherit* __this){
delete __this;
}



_testDataFieldInherit::_testDataFieldInherit(): __vptr(&__vtable){

}

void _testI1::__delete(_testI1* __this){
delete __this;
}

void _testI1::test(testI1 __this){
std::cout << ({__rt::literal("test1.test(): Should print original name:");}) << std::endl;
std::cout << ({(__rt::literal("\t") + __this->name);}) << std::endl;

}

void _testI1::test2(testI1 __this){
std::cout << ({__rt::literal("test1.test2(): Should print original name:");}) << std::endl;
std::cout << ({(__rt::literal("\t") + __this->name);}) << std::endl;

}

_testI1::_testI1(): __vptr(&__vtable){
	this
}

void _testI2::__delete(_testI2* __this){
delete __this;
}

void _testI2::test(testI2 __this){
std::cout << ({__rt::literal("test2.test(): Should print 69:");}) << std::endl;
std::cout << ({(__rt::literal("\t") + __this->name);}) << std::endl;

}

void _testI2::test2(testI2 __this){
std::cout << ({__rt::literal("test2.test2(): Should print original name:");}) << std::endl;
std::cout << ({(__rt::literal("\t") + __this->$name);}) << std::endl;

}

_testI2::_testI2(): __vptr(&__vtable){

}

void _testI3::__delete(_testI3* __this){
delete __this;
}

void _testI3::test(testI3 __this){
std::cout << ({(__rt::literal("Should print 69: ") + __this->$name);}) << std::endl;
std::cout << ({(__rt::literal("Shoudld print c: ") + __this->name);}) << std::endl;

}

_testI3::_testI3(): __vptr(&__vtable){

}

void _testI4::__delete(_testI4* __this){
delete __this;
}

void _testI4::test(testI4 __this){
std::cout << ({(__rt::literal("Should print c: ") + __this->name);}) << std::endl;
std::cout << ({(__rt::literal("Shoudld print c: ") + __this->name);}) << std::endl;

}

_testI4::_testI4(): __vptr(&__vtable){

}

int main(int argc, const char* argv[]) {
/*
__rt::Ptr<__rt::Array<String> > args = new __rt::Array<String>(argc-1);
for(int i = 1; i < argc; i++){
(*args)[i-1] = argv[i];
}
*/
testI1 t1 = new _testI1();
t1->__vptr->test(t1);
t1->__vptr->test2(t1);
testI2 t2 = new _testI2();
t2->__vptr->test(t2);
t2->__vptr->test2(t2);
testI3 t3 = new _testI3();
t3->__vptr->test(t3);
testI4 t4 = new _testI4();
t4->__vptr->test(t4);
return 0;
}

_testDataFieldInherit_VT _testDataFieldInherit::__vtable;

Class _testDataFieldInherit::__class() {
 static Class k = new java::lang::__Class(__rt::literal("testDataFieldInherit"), java::lang::__Object::__class());
return k;
}

_testI1_VT _testI1::__vtable;

Class _testI1::__class() {
 static Class k = new java::lang::__Class(__rt::literal("testI1"), java::lang::__Object::__class());
return k;
}

_testI2_VT _testI2::__vtable;

Class _testI2::__class() {
 static Class k = new java::lang::__Class(__rt::literal("testI2"), java::lang::__Object::__class());
return k;
}

_testI3_VT _testI3::__vtable;

Class _testI3::__class() {
 static Class k = new java::lang::__Class(__rt::literal("testI3"), java::lang::__Object::__class());
return k;
}

_testI4_VT _testI4::__vtable;

Class _testI4::__class() {
 static Class k = new java::lang::__Class(__rt::literal("testI4"), java::lang::__Object::__class());
return k;
}

